### PR TITLE
api: add interface for hot_region info reqeust

### PR DESCRIPTION
**Pull request by qidi1**
…onfg

Signed-off-by: qidi1 <1083369179@qq.com>

### What problem does this PR solve?
origin pr is here https://github.com/tikv/pd/pull/3989
**This PR is oversize and splited into three small one:**

1. https://github.com/tikv/pd/pull/4019
2. https://github.com/tikv/pd/pull/4020
2. https://github.com/tikv/pd/pull/4021
3. https://github.com/tikv/pd/pull/4103
4. https://github.com/tikv/pd/pull/4121

Proposal: [pingcap/tidb#27487](https://github.com/pingcap/tidb/pull/27487)
Associated PR:[tikv/tidb#27224](https://github.com/pingcap/tidb/pull/27224)
Associated Issue: [pingcap/tidb#25281](https://github.com/pingcap/tidb/issues/25281)
1. [pingcap/tidb#27373](https://github.com/pingcap/tidb/pull/27373)
2. [pingcap/tidb#27374](https://github.com/pingcap/tidb/pull/27374)
3. [pingcap/tidb#27374](https://github.com/pingcap/tidb/pull/27375)

### What is changed and how it works?
add api for hot_region_request
### Check List

Tests

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Code changes

- [ ] Has configuration change
- [x] Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- [ ] Has persistent data change

Side effects

- [ ] Possible performance regression
- [ ] Increased code complexity
- [ ] Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tidb-ansible`](https://github.com/pingcap/tidb-ansible):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/api/hot_status.go [hunk 1]
```diff
@@ -141,3 +143,30 @@ func (h *hotStatusHandler) GetHotStores(w http.ResponseWriter, r *http.Request)
...
+	data, err := io.ReadAll(r.Body)
+	r.Body.Close()
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	historyHotRegionsRequest := &statistics.HistoryHotRegionsRequest{}
+	err = json.Unmarshal(data, historyHotRegionsRequest)
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
```
Comment (disksing) [inline server/api/hot_status.go:163]: better return 400.

File: server/api/hot_status.go [hunk 2]
```diff
@@ -141,3 +143,30 @@ func (h *hotStatusHandler) GetHotStores(w http.ResponseWriter, r *http.Request)
 	}
 	h.rd.JSON(w, http.StatusOK, stats)
 }
+
+// @Tags hotspot
+// @Summary List the history hot regions.
+// @Accept json
+// @Produce json
+// @Success 200 {object} statistics.HistoryHotRegions
+// @Router /hotspot/regions/history [post]
```
Comment (disksing) [inline server/api/hot_status.go:152]: `get` should be more suitable.

File: server/api/router.go [hunk 3]
```diff
@@ -140,6 +140,7 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	apiRouter.HandleFunc("/hotspot/regions/write", hotStatusHandler.GetHotWriteRegions).Methods("GET")
 	apiRouter.HandleFunc("/hotspot/regions/read", hotStatusHandler.GetHotReadRegions).Methods("GET")
 	apiRouter.HandleFunc("/hotspot/stores", hotStatusHandler.GetHotStores).Methods("GET")
+	apiRouter.HandleFunc("/hotspot/regions/history", hotStatusHandler.GetHistoryHotRegions).Methods("POST")
```
Comment (disksing) [inline server/api/router.go:143]: consider use `GET`

File: server/handler.go [hunk 4]
```diff
@@ -898,6 +898,66 @@ func (h *Handler) SetStoreLimitTTL(data string, value float64, ttl time.Duration
...
+		if len(regionSet) != 0 && !regionSet[next.RegionID] {
+			continue
+		}
+		if len(storeSet) != 0 && !storeSet[next.StoreID] {
+			continue
+		}
+		if len(peerSet) != 0 && !peerSet[next.PeerID] {
+			continue
+		}
+		if len(roleSet) != 0 && !roleSet[next.IsLeader] {
```
Comment (disksing) [inline server/handler.go:939]: I'm a little confused. Does `next.IsLeader` have a type `int64`?

File: server/handler.go [hunk 5]
```diff
@@ -898,6 +898,66 @@ func (h *Handler) SetStoreLimitTTL(data string, value float64, ttl time.Duration
...
+// GetAllRequestHistroyHotRegion get historyHotRegions as request.
+func (h *Handler) GetAllRequestHistroyHotRegion(request *statistics.HistoryHotRegionsRequest) (*statistics.HistoryHotRegions, error) {
+	var hotRegionTypes []string
+	if len(request.HotRegionTypes) != 0 {
+		hotRegionTypes = request.HotRegionTypes
+	} else {
+		hotRegionTypes = cluster.HotRegionTypes
+	}
+	iter := h.s.hotRegionStorage.NewIterator(hotRegionTypes, request.StartTime, request.EndTime)
+	results := make([]*statistics.HistoryHotRegion, 0)
```
Comment (disksing) [inline server/handler.go:910]:
```suggestion
	var results []*statistics.HistoryHotRegion
```

File: server/handler.go [hunk 6]
```diff
@@ -898,6 +898,66 @@ func (h *Handler) SetStoreLimitTTL(data string, value float64, ttl time.Duration
...
 }
 
+// GetAllRequestHistroyHotRegion get historyHotRegions as request.
+func (h *Handler) GetAllRequestHistroyHotRegion(request *statistics.HistoryHotRegionsRequest) (*statistics.HistoryHotRegions, error) {
+	var hotRegionTypes []string
+	if len(request.HotRegionTypes) != 0 {
+		hotRegionTypes = request.HotRegionTypes
+	} else {
+		hotRegionTypes = cluster.HotRegionTypes
+	}
```
Comment (disksing) [inline server/handler.go:908]:
```suggestion
	hotRegionTypes := cluster.HotRegionTypes
	if len(request.HotRegionTypes) != 0 {
		hotRegionTypes = request.HotRegionTypes
	}
```

File: server/api/hot_status.go (hunk 2 above)
Comment (qidi1) [reply server/api/hot_status.go:174]: Too many parameters are needed, it is troublesome to switch to Get form

File: server/api/hot_status.go [hunk 7]
```diff
@@ -38,6 +41,27 @@ type HotStoreStats struct {
...
+// HistoryHotRegionsRequest wrap request condition from tidb.
+// it is request from tidb
+type HistoryHotRegionsRequest struct {
+	StartTime int64    `json:"start_time,omitempty"`
+	EndTime   int64    `json:"end_time,omitempty"`
+	RegionIDs []uint64 `json:"region_ids,omitempty"`
+	StoreIDs  []uint64 `json:"store_ids,omitempty"`
+	PeerIDs   []uint64 `json:"peer_ids,omitempty"`
+	//0 means not leader,1 means leader
+	Roles          []int64  `json:"is_leader,omitempy"`
```
Comment (rleungx) [inline server/api/hot_status.go:53]: Use bool and change the variable name?

File: server/handler.go [hunk 8]
```diff
@@ -898,6 +898,14 @@ func (h *Handler) SetStoreLimitTTL(data string, value float64, ttl time.Duration
 	}, ttl)
 }
 
+// GetHistoryHotRegionIter return a iter which iter all qualified item .
+func (h *Handler) GetHistoryHotRegionIter(hotRegionTypes []string,
+	StartTime, EndTime int64) cluster.HotRegionStorageIterator {
+	iter := h.s.hotRegionStorage.NewIterator(hotRegionTypes, StartTime, EndTime)
+	return iter
+
```
Comment (rleungx) [inline server/handler.go:906]: remove the empty line

File: server/api/hot_status.go (hunk 2 above)
Comment (disksing) [reply server/api/hot_status.go:174]: I think GET can have body too.

File: server/api/hot_status.go (hunk 7 above)
Comment (qidi1) [reply server/api/hot_status.go:53]: In TiDB,bool use int64

File: server/api/hot_status_test.go [hunk 9]
```diff
@@ -306,3 +310,29 @@ func writeToDB(c *C, kv *kv.LeveldbKV, hotRegions []*statistics.HistoryHotRegion
 	}
 	kv.Write(batch, nil)
 }
+
+func getJSON(client *http.Client, url string, data []byte, checkOpts ...func([]byte, int)) error {
```
Comment (rleungx) [inline server/api/hot_status_test.go:314]: move it to `server/api/util.go`

File: server/api/hot_status.go [hunk 10]
```diff
@@ -141,3 +157,78 @@ func (h *hotStatusHandler) GetHotStores(w http.ResponseWriter, r *http.Request)
 	}
 	h.rd.JSON(w, http.StatusOK, stats)
 }
+
+// @Tags hotspot
+// @Summary List the history hot regions.
+// @Accept json
+// @Produce json
+// @Success 200 {object} statistics.HistoryHotRegions
+// @Failure 400 {string} string "The input is invalid."
```
Comment (rleungx) [inline server/api/hot_status.go:166]: need 500

File: server/api/hot_status.go (hunk 7 above)
Comment (rleungx) [reply server/api/hot_status.go:53]: BTW, it's better to be consistent with the JSON tag.

File: server/api/hot_status_test.go [hunk 11]
```diff
@@ -48,3 +54,137 @@ func (s testHotStatusSuite) TestGetHotStore(c *C) {
...
+	storage := s.svr.GetHistoryHotRegionStorage()
+	now := time.Now()
+	hotRegions := []*core.HistoryHotRegion{
+		{
+			RegionID:      1,
+			StoreID:       1,
+			PeerID:        1,
+			IsLeader:      false,
+<<<<<<< HEAD
+=======
```
Comment (nolouch) [inline server/api/hot_status_test.go:112]: need to resolve the conflict.

**Comment by qidi1** [comment]
@nolouch All Test pass

**Review by nolouch — APPROVED** [review]
lgtm

**Comment by nolouch** [comment]
PTAL @rleungx @disksing

File: server/api/util.go [hunk 12]
```diff
@@ -91,24 +91,20 @@ func extractJSON(resp *http.Response, data interface{}) error {
 }
 
 func postJSON(client *http.Client, url string, data []byte, checkOpts ...func([]byte, int)) error {
-	resp, err := client.Post(url, "application/json", bytes.NewBuffer(data))
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(data))
 	if err != nil {
-		return errors.WithStack(err)
+		return nil
```
Comment (rleungx) [inline server/api/util.go:96]: Why return nil?

File: server/api/util.go [hunk 13]
```diff
@@ -91,24 +91,20 @@ func extractJSON(resp *http.Response, data interface{}) error {
...
 	if err != nil {
-		return err
-	}
-
-	if resp.StatusCode != http.StatusOK {
-		return errors.New(string(res))
-	}
-	for _, opt := range checkOpts {
-		opt(res, resp.StatusCode)
+		return nil
```
Comment (rleungx) [inline server/api/util.go:105]: ditto

File: server/api/router.go [hunk 14]
```diff
@@ -154,6 +154,7 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	apiRouter.HandleFunc("/hotspot/regions/write", hotStatusHandler.GetHotWriteRegions).Methods("GET")
 	apiRouter.HandleFunc("/hotspot/regions/read", hotStatusHandler.GetHotReadRegions).Methods("GET")
 	apiRouter.HandleFunc("/hotspot/stores", hotStatusHandler.GetHotStores).Methods("GET")
+	apiRouter.HandleFunc("/hotspot/regions/history", hotStatusHandler.GetHistoryHotRegions).Methods("GET")
```
Comment (rleungx) [inline server/api/router.go:157]: prefer to exchange with line 156

File: server/api/hot_status_test.go [hunk 15]
```diff
@@ -49,3 +55,143 @@ func (s testHotStatusSuite) TestGetHotStore(c *C) {
...
+
+func writeToDB(c *C, kv *kv.LeveldbKV, hotRegions []*core.HistoryHotRegion) {
+	batch := new(leveldb.Batch)
+	for _, region := range hotRegions {
+		key := core.HotRegionStorePath(region.HotRegionType, region.UpdateTime, region.RegionID)
+		value, err := json.Marshal(region)
+		c.Assert(err, IsNil)
+		batch.Put([]byte(key), value)
+	}
+	kv.Write(batch, nil)
```
Comment (rleungx) [inline server/api/hot_status_test.go:196]: check error?

File: server/api/hot_status.go [hunk 16]
```diff
@@ -153,3 +168,82 @@ func (h *hotStatusHandler) GetHotStores(w http.ResponseWriter, r *http.Request)
...
 	h.rd.JSON(w, http.StatusOK, stats)
 }
+
+// @Tags hotspot
+// @Summary List the history hot regions.
+// @Accept json
+// @Produce json
+// @Success 200 {object} core.HistoryHotRegions
+// @Failure 400 {string} string "The input is invalid."
+// @Failure 500 {string} string "server error:can`t get history hot region info"
```
Comment (rleungx) [inline server/api/hot_status.go:178]:
```suggestion
// @Failure 500 {string} string "PD server failed to proceed the request."
```

**Review by disksing — APPROVED** [review]

File: server/api/hot_status.go (hunk 16 above)
Comment (qidi1) [reply server/api/hot_status.go]: apply to code

File: server/api/hot_status_test.go (hunk 15 above)
Comment (qidi1) [reply server/api/hot_status_test.go:199]: move to test to check

File: server/api/util.go (hunk 13 above)
Comment (qidi1) [reply server/api/util.go]: change to 156

File: server/api/util.go (hunk 12 above)
Comment (qidi1) [reply server/api/util.go]: Accidentally made a mistake

**Pull request merged**